org/mozilla/jss/crypto/KBKDFCounterParams.java

```diff
@@ -0,0 +1,204 @@
+    }
+    public KBKDFCounterParams(long prf, SecretKey prfKey, long algo, int keySize) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+    }
+    public KBKDFCounterParams(PKCS11Algorithm prf, SecretKey prfKey, PKCS11Algorithm algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+    public KBKDFCounterParams(DigestAlgorithm prf, SecretKey prfKey, PKCS11Algorithm algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+    public KBKDFCounterParams(long prf, SecretKey prfKey, PKCS11Algorithm algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+    public KBKDFCounterParams(PKCS11Algorithm prf, SecretKey prfKey, long algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+    public KBKDFCounterParams(DigestAlgorithm prf, SecretKey prfKey, long algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+    public KBKDFCounterParams(long prf, SecretKey prfKey, long algo, int keySize, KBKDFDataParameter[] params) throws IllegalArgumentException, InvalidKeyException {
+        setPRF(prf);
+        setPRFKey(prfKey);
+        setDerivedKeyAlgorithm(algo);
+        setKeySize(keySize);
+        setParameters(params);
+    }
+
+    protected void validateParameters() throws IllegalArgumentException {
+        if (prfKey == null) {
+            String msg = "Required parameter KDF key was never set.";
+            throw new IllegalArgumentException(msg);
+        }
+
+        if (prf != PKCS11Constants.CKM_SHA_1_HMAC &&
+                prf != PKCS11Constants.CKM_SHA256_HMAC &&
+                prf != PKCS11Constants.CKM_SHA384_HMAC &&
+                prf != PKCS11Constants.CKM_SHA512_HMAC &&
+                prf != PKCS11Constants.CKM_AES_CMAC) {
+            String msg = "Expected Pseudo-Random Function to be SHA1-HMAC, ";
+            msg += "SHA2-HMAC, or AES-CMAC, but got unknown PKCS#11 ";
```

Comment:
Here is the error message exactly correct? For instance it doesn't list all the HMACs it is checking for above..

Reply:
A [`SHA2`](https://en.wikipedia.org/wiki/SHA-2)-HMAC is any of `CKM_SHA256_HMAC`, `CKM_SHA384_HMAC`, or `CKM_SHA512_HMAC`.

(Technically it also includes `CKM_SHA224_HMAC`, but JSS doesn't support that anywhere so its ommission shouldn't be surprising to the caller.)

Reply:
OK, I thought maybe there was shorthand going on there, just double checking..